presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java

```diff
@@ -255,6 +256,26 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Set<Symbol>> con
                     node.getDistributionType());
         }
 
+        @Override
+        public PlanNode visitSpatialJoin(SpatialJoinNode node, RewriteContext<Set<Symbol>> context)
+        {
+            Set<Symbol> expectedFilterInputs = ImmutableSet.<Symbol>builder()
+                    .addAll(SymbolsExtractor.extractUnique(node.getFilter()))
+                    .addAll(context.get())
+                    .build();
+
+            PlanNode left = context.rewrite(node.getLeft(), expectedFilterInputs);
+            PlanNode right = context.rewrite(node.getRight(), expectedFilterInputs);
+
+            Set<Symbol> seenSymbol = new HashSet<>();
+            List<Symbol> outputSymbols = node.getOutputSymbols().stream()
+                    .filter(context.get()::contains)
+                    .filter(seenSymbol::add)
```

Comment:
You can use `distinct`

Reply:
It's using `distinct()` right above the `filter(seenSymbol::add)`, which makes the call to `filter(seenSymbol::add)` and the `seenSymbol` set unnecessary. As a side note, try to avoid having side-effects when using streams (except for the final `forEach`). It makes it hard to read and reason about the code, and it's error prone.